scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+# mypy: disable-error-code="attr-defined"
+import numpy as np
+from scipy import special
+from scipy.optimize import OptimizeResult
+from scipy.optimize._zeros_py import (_scalar_optimization_initialize,
+                                      _scalar_optimization_loop,
+                                      _ECONVERGED, _ESIGNERR, _ECONVERR,  # noqa
+                                      _EVALUEERR, _ECALLBACK, _EINPROGRESS)  # noqa
+
+# todo:
+#  figure out warning situation
+#  address https://github.com/scipy/scipy/pull/18650#discussion_r1233032521
+#  without `minweight`, we are also suppressing infinities within the interval.
+#    Is that OK? If so, we can probably get rid of `status=3`.
+#  Add heuristic to stop when improvement is too slow / antithrashing
+#  support singularities? interval subdivision? this feature will be added
+#    eventually, but do we adjust the interface now?
+#  When doing log-integration, should the tolerances control the error of the
+#    log-integral or the error of the integral?  The trouble is that `log`
+#    inherently looses some precision so it may not be possible to refine
+#    the integral further. Example: 7th moment of stats.f(15, 20)
+#  respect function evaluation limit?
```

Comment:
I used to accept an argument to control the maximum number of function evaluations. For simplicity, I took it out when vectorizing, but I could do it in a follow-up if we think it's important. (I considered building it into the scalar optimization algorithm framework, but it didn't seem so imporant: for all of the other algorithms so far, the number if function evaluations is very easily related to the number of iterations. I only bring it up here because the relationship is more complex for this algorithm, and the number of function evaluations can be orders of magnitude different from the number of iterations.)